### PR TITLE
C++ sgd optimizer

### PR DESCRIPTION
**Pull request by piiswrong**
This is faster and doesn't allocate temporary memory. Registered as ccsgd. Should we make this the default?

### CODE REVIEW

**Comment by tqchen** [comment]
LGTM, I am still debating on towards what level can these operations be exposed, should it be a series of mshadow operations(so no allocation was involved), or pushed as an complete optimizer interface, can you open an issue on this? Some speed comparison on benchmarks will also be helpful

**Comment by tqchen** [comment]
@jermainewang @pluskid

File: src/operator/cudnn_convolution-inl.h [hunk 1]
```diff
@@ -211,7 +211,7 @@ class CuDNNConvolutionOp : public Operator {
                filter_desc_,
                conv_desc_,
                out_desc_,
-               CUDNN_CONVOLUTION_FWD_PREFER_FASTEST,
+               CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT,
```
Comment (piiswrong) [inline src/operator/cudnn_convolution-inl.h]: @antinucleon I suspect this is what caused the NeuralStyle error when input is too large. Please check if this fixes it

**Comment by pluskid** [comment]
Maybe it makes sense to expose a simplified symbolic computing interface to allow the user specify the computation graph of an optimizer conveniently?

**Comment by piiswrong** [comment]
My reason for using C++ optimizer is to avoid temp memory. That kind of defeats the purpose

**Comment by piiswrong** [comment]
@tqchen This PR also include some bug fixes. Could you review and merge soon

**Comment by tqchen** [comment]
Will do a review t, can you isolate the bugfix first with another PR?

**Pull request merged**